pkg/ottl/ottlfuncs/func_replace_all_patterns.go

```diff
@@ -44,6 +45,41 @@ func createReplaceAllPatternsFunction[K any](_ ottl.FunctionContext, oArgs ottl.
 	return replaceAllPatterns(args.Target, args.Mode, args.RegexPattern, args.Replacement, args.Function)
 }
 
+func applyReplaceAllFunction[K any](ctx context.Context, tCtx K, compiledPattern *regexp.Regexp, fn ottl.Optional[ottl.FunctionGetter[K]], originalValStr string, replacementVal string) (string, error) {
+	var updatedString string
+	updatedString = originalValStr
+	submatches := compiledPattern.FindAllStringSubmatchIndex(updatedString, -1)
+	for _, submatch := range submatches {
+		fullMatch := originalValStr[submatch[0]:submatch[1]]
+		result := compiledPattern.ExpandString([]byte{}, replacementVal, originalValStr, submatch)
+		fnVal := fn.Get()
+		replaceValGetter := ottl.StandardStringGetter[K]{
+			Getter: func(context.Context, K) (any, error) {
+				return string(result), nil
+			},
+		}
+		replacementExpr, errNew := fnVal.Get(&replaceAllPatternFuncArgs[K]{Input: replaceValGetter})
+		if errNew != nil {
+			return "", errNew
+		}
+		replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
+		if errNew != nil {
+			return "", errNew
+		}
+		replacementValStr, ok := replacementValRaw.(string)
+		if !ok {
+			return "", fmt.Errorf("the replacement value must be a string")
+		}
+		if len(submatch) > 2 {
+			updatedString = strings.ReplaceAll(updatedString, fullMatch, replacementValStr)
+		}
+		if !IsGroupNumString(replacementVal) {
```

Comment:
this covers a scenario where there are no submatches and the user provides a regular replacement string with a hash function. In this case we, just ignore the hash function.

Reply:
I understand the reasoning for that, but in my opinion we should always run the optional function if the user passes one in. I think it would be unexpected for the function to not be run if it is provided.

Reply:
Looking at it again, I think there's actually good reason to provide this outside consistency: the replacement value is a StringGetter, so a user may want to e.g. replace all matches with a hashed value from another part of the telemetry item. We should probably remove both checks and just always run the replacement regardless of whether any submatches were found or specified.

Reply:
makes sense, I made the update for consistent behavior.

Reply:
the upload artifact github action seems to be failing
`Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: an artifact with this name already exists on the workflow run`